File: dotnet-interop-managed-lib/Computer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LHM = LibreHardwareMonitor.Hardware;

namespace dotnet_interop_net_lib
	{
	class InvalidIdentifierException : Exception { public InvalidIdentifierException(string identifier) : base("Identifier [" + identifier + "] is not valid.") { } }
	class InvalidIdentifierInSensorException : Exception { public InvalidIdentifierInSensorException(string identifier, string sensor_id) : base("Trying to search identifier [" + identifier + "] inside sensor." + sensor_id + ".") { } }
	class IdentifierNotFoundException : Exception { public IdentifierNotFoundException(string identifier, string parent_id) : base("Cannot find identifier [" + identifier + "] inside " + parent_id + ".") { } }
	class SensorNotFoundException : Exception { public SensorNotFoundException(string identifier) : base("Cannot find sensor [" + identifier + "]. Is it an hardware?") { } }
	class HardwareNotFoundException : Exception { public HardwareNotFoundException(string identifier) : base("Cannot find hardware [" + identifier + "]. Is it a sensor?") { } }
	class IncompatibleSensorsException : Exception { public IncompatibleSensorsException(Sensor a, Sensor b) : base("Sensors [" + a.identifier + "] and [" + b.identifier + "] are not compatible.") { } }

	public abstract class ComputerComponent
		{
		public abstract string name { get; }
		public abstract string identifier { get; }
		public abstract string to_json(int indent = 0);
		}

	public abstract class Sensor : ComputerComponent
		{
		abstract public LHM.SensorType type { get; }
		public string value_string
			{
			get
				{
				switch (type)
					{
					case LHM.SensorType.Data: return Utils.ValuesConversions.dataGB(value);
					case LHM.SensorType.SmallData: return Utils.ValuesConversions.dataMB(value);
					case LHM.SensorType.Fan: return Utils.ValuesConversions.fan(value);
					case LHM.SensorType.Level:
					case LHM.SensorType.Control:
					case LHM.SensorType.Load: return Utils.ValuesConversions.load(value);
					case LHM.SensorType.Temperature: return Utils.ValuesConversions.temperature(value);
					default: return "TODO" + (value.HasValue ? value.Value.ToString() : "");
					}
				}
			}

		public abstract float? value { get; }
		public abstract float? min { get; }
		public abstract float? max { get; }

		public override string to_json(int indent = 0)
			{
			return Utils.s.indent(indent) + "{ " +
			"\"name\": " + Utils.s.uniform_length("\"" + name + "\"", 30) +
			", \"id\": " + Utils.s.uniform_length("\"" + identifier + "\"", 30) +
			", \"type\": " + Utils.s.uniform_length("\"" + type + "\"", 15) + "" +
			", \"value\": " + Utils.s.uniform_length("\"" + (value.HasValue ? value.Value.ToString() : "N/A") + "\"", 15) + "}";
			}
		}

	class SensorCombiner : Sensor
		{
		private Sensor a = null, b = null;
		public SensorCombiner(Sensor a, Sensor b)
			{
			if (a.type != b.type &&
				!((a.type == LHM.SensorType.Data || a.type == LHM.SensorType.SmallData) && (b.type == LHM.SensorType.Data || b.type == LHM.SensorType.SmallData)))
				{ throw new IncompatibleSensorsException(a, b); }
			this.a = a; this.b = b;
			}

		public override LHM.SensorType type { get { return a.type; } }

		public override string name { get { return a.name + " + " + b.name; } }
		public override string identifier { get { return "{" + a.identifier + "+" + b.identifier + "}"; } }

		public override float? value { get { return a.value + b.value; } }
		public override float? min { get { return a.min + b.min; } }
		public override float? max { get { return a.max + b.max; } }
		}

	class SensorDirect : Sensor
		{
		private LHM.ISensor _this = null;
		public SensorDirect(LHM.ISensor lhm_sensor) { this._this = lhm_sensor; }

		public override string name { get { return _this.Name; } }
		public override string identifier { get { return _this.Identifier.ToString(); } }
		public int index { get { return _this.Index; } }

		public override LHM.SensorType type { get { return _this.SensorType; } }

		public override float? value { get { return _this.Value; } }
		public override float? min { get { return _this.Min; } }
		public override float? max { get { return _this.Max; } }
		}

	public class Hardware : ComputerComponent
		{
		private LHM.IHardware _this = null;
		public Hardware(LHM.IHardware lhm_hardware) { this._this = lhm_hardware; }

		public override string name { get { return _this.Name; } }
		public override string identifier { get { return _this.Identifier.ToString(); } }

		public LHM.HardwareType type { get { return _this.HardwareType; } }

		public int sub_hardware_size { get { return _this.SubHardware.Length; } }
		public Utils.IndexedPropertyReadOnly<int, Hardware> sub_hardware
			{ get { return new Utils.IndexedPropertyReadOnly<int, Hardware>(index => { return new Hardware(_this.SubHardware[index]); }); } }

		public int sensors_size { get { return _this.Sensors.Length; } }
		public Utils.IndexedPropertyReadOnly<int, Sensor> sensors
			{ get { return new Utils.IndexedPropertyReadOnly<int, Sensor>(index => { return new SensorDirect(_this.Sensors[index]); }); } }

		public void update() { _this.Update(); foreach (LHM.IHardware lhm_hardware in _this.SubHardware) { new Hardware(lhm_hardware).update(); } }

		public override string to_json(int indent = 0)
			{
			update();
			string ret = Utils.s.indent(indent) + "{ " +
			"\"name\": \"" + name + "\"" +
			", \"id\": \"" + identifier + "\"" +
			", \"type\": \"" + type + "\"";
			if (_this.Sensors.Length > 0)
				{
				ret += ",\n" +
				Utils.s.indent(indent) + "\"sensor\": \n" +
				Utils.s.indent(indent + 1) + "[\n";
				for (int i = 0; i < sensors_size - 1; i++) { ret += sensors[i].to_json(indent + 2) + ",\n"; }
				ret += sensors[sensors_size - 1].to_json(indent + 2) + "\n" +
				Utils.s.indent(indent + 1) + "]";
				}
			if (_this.SubHardware.Length > 0)
				{
				ret += ",\n" +
				Utils.s.indent(indent) + "\"hardware\": \n" +
				Utils.s.indent(indent + 1) + "[\n";
				for (int i = 0; i < sub_hardware_size - 1; i++) { ret += sub_hardware[i].to_json(indent + 2) + ",\n"; }
				ret += sub_hardware[sub_hardware_size - 1].to_json(indent + 2) + "\n" +
				Utils.s.indent(indent + 1) + "]";
				}
			ret += "\n" + Utils.s.indent(indent) + "}";
			return ret;
			}
		}

	public class Computer : IDisposable
		{
		private LHM.Computer _this = null;

		public Computer()
			{
			_this = new LHM.Computer() { IsCpuEnabled = true, IsGpuEnabled = true, IsMemoryEnabled = true, IsMotherboardEnabled = true, IsControllerEnabled = true, IsStorageEnabled = true, IsNetworkEnabled = true };
			_this.Open();
			}

		public void update() { foreach (LHM.IHardware lhm_hardware in _this.Hardware) { new Hardware(lhm_hardware).update(); } }

		private Sensor get_sensor_in_hardware(string s, LHM.IHardware hardware)
			{
			Sensor found = null;
			foreach (var sensor in hardware.Sensors) { if (sensor.Identifier.ToString().Equals(s)) { return new SensorDirect(sensor); } }
			foreach (var sub_hardware in hardware.SubHardware) { found = get_sensor_in_hardware(s, sub_hardware); if (found != null) { return found; } }
			return null;
			}
		private Hardware get_hardware_in_hardware(string s, LHM.IHardware hardware)
			{
			if (hardware.Identifier.ToString().Equals(s)) { return new Hardware(hardware); }
			Hardware found = null;
			foreach (var sub_hardware in hardware.SubHardware) { found = get_hardware_in_hardware(s, sub_hardware); if (found != null) { return found; } }
			return null;
			}
		private ComputerComponent get_in_hardware(string s, LHM.IHardware hardware)
			{
			if (hardware.Identifier.ToString().Equals(s)) { return new Hardware(hardware); }
			ComputerComponent found = null;
			foreach (var sensor in hardware.Sensors) { if (sensor.Identifier.ToString().Equals(s)) { return new SensorDirect(sensor); } }
			foreach (var sub_hardware in hardware.SubHardware) { found = get_in_hardware(s, sub_hardware); if (found != null) { return found; } }
			return null;
			}

		public Sensor get_sensor(string s)
			{
			Sensor found = null;
			foreach (var hardware in _this.Hardware) { found = get_sensor_in_hardware(s, hardware); if (found != null) { return found; } }
			throw new SensorNotFoundException(s);
			}
		public Hardware get_hardware(string s)
			{
			Hardware found = null;
			foreach (var hardware in _this.Hardware) { found = get_hardware_in_hardware(s, hardware); if (found != null) { return found; } }
			throw new HardwareNotFoundException(s);
			}
		public ComputerComponent get(string s)
			{
			ComputerComponent found = null;
			foreach (var hardware in _this.Hardware) { found = get_in_hardware(s, hardware); if (found != null) { return found; } }
			throw new HardwareNotFoundException(s);
			}

		public string to_json(int indent = 0)
			{
			string ret = Utils.s.indent(indent) + "{ " +
			Utils.s.indent(indent) + "\"hardware\": \n" +
			Utils.s.indent(indent + 1) + "[\n";
			if (hardware_size > 0)
				{
				for (int i = 0; i < hardware_size - 1; i++) { ret += hardware[i].to_json(indent + 2) + ",\n"; }
				ret += hardware[hardware_size - 1].to_json(indent + 2) + "\n";
				}
			ret += Utils.s.indent(indent + 1) + "]\n" +
			Utils.s.indent(indent) + "}";
			return ret;
			}


		public int hardware_size { get { return _this.Hardware.Count; } }
		public Utils.IndexedPropertyReadOnly<int, Hardware> hardware
			{ get { return new Utils.IndexedPropertyReadOnly<int, Hardware>(index => { return new Hardware(_this.Hardware[index]); }); } }


		public void WriteToConsole(string message)
			{
			Console.WriteLine(message);
			}

		// DESTRUCTOR
		private bool disposed = false;
		~Computer() => Dispose(false);
		protected virtual void Dispose(bool disposing)
			{
			if (disposed) { return; }
			if (disposing) { }
			_this?.Close(); _this = null;

			disposed = true;
			}
		public void Dispose() { Dispose(disposing: true); GC.SuppressFinalize(this); }
		}
	}

File: dotnet-interop-clr/Computer.cpp
#include "Computer.h"
#include <gcroot.h>
#include <msclr/marshal_cppstd.h>
#include <optional>
namespace LHM = LibreHardwareMonitor::Hardware;

sensor_type sensor_type_conversion(LHM::SensorType type)
	{
	switch (type)
		{
			case LibreHardwareMonitor::Hardware::SensorType::Voltage:  	  return sensor_type::voltage;
			case LibreHardwareMonitor::Hardware::SensorType::Clock:		  return sensor_type::clock;
			case LibreHardwareMonitor::Hardware::SensorType::Temperature: return sensor_type::temperature;
			case LibreHardwareMonitor::Hardware::SensorType::Load:		  return sensor_type::load;
			case LibreHardwareMonitor::Hardware::SensorType::Frequency:	  return sensor_type::frequency;
			case LibreHardwareMonitor::Hardware::SensorType::Fan:		  return sensor_type::fan;
			case LibreHardwareMonitor::Hardware::SensorType::Flow:		  return sensor_type::flow;
			case LibreHardwareMonitor::Hardware::SensorType::Control:	  return sensor_type::control;
			case LibreHardwareMonitor::Hardware::SensorType::Level:		  return sensor_type::level;
			case LibreHardwareMonitor::Hardware::SensorType::Factor:	  return sensor_type::factor;
			case LibreHardwareMonitor::Hardware::SensorType::Power:		  return sensor_type::power;
			case LibreHardwareMonitor::Hardware::SensorType::Data:		  return sensor_type::data;
			case LibreHardwareMonitor::Hardware::SensorType::SmallData:	  return sensor_type::smalldata;
			case LibreHardwareMonitor::Hardware::SensorType::Throughput:  return sensor_type::throughput;
			default: 													  return sensor_type::unknown;
		}
	}
std::string sensor_type_to_string(sensor_type type)
	{
	switch (type)
		{
			case sensor_type::voltage:			return "voltage";
			case sensor_type::clock:			return "clock";
			case sensor_type::temperature:		return "temperature";
			case sensor_type::load:				return "load";
			case sensor_type::frequency:		return "frequency";
			case sensor_type::fan:				return "fan";
			case sensor_type::flow:				return "flow";
			case sensor_type::control:			return "control";
			case sensor_type::level:			return "level";
			case sensor_type::factor:			return "factor";
			case sensor_type::power:			return "power";
			case sensor_type::data:				return "data";
			case sensor_type::smalldata:		return "smalldata";
			case sensor_type::throughput:		return "throughput";
			case sensor_type::unknown: default: return "Unknown";
		}
	}

std::string indent(uint8_t depth)
	{
	std::string ret = "";
	for (uint8_t i = 0; i < depth; i++) { ret += "    "; }
	return ret;
	}

LHM::ISensor^ get_sensor(LHM::IHardware^ hardware, System::String^ identifier)
	{
	for each (auto h in hardware->SubHardware)
		{
		LHM::ISensor^ tmp = get_sensor(h, identifier);
		if (tmp != nullptr) 
			{
			tmp->ValuesTimeWindow = System::TimeSpan::Zero;
			return tmp; 
			}
		}
	for each (auto s in hardware->Sensors)
		{
		if (s->Identifier->ToString()->Equals(identifier)) { return s; }
		}
	return nullptr;
	}

LHM::IHardware^ get_hardware(LHM::IHardware^ hardware, System::String^ identifier)
	{
	for each (auto h in hardware->SubHardware)
		{
		LHM::IHardware^ tmp = get_hardware(h, identifier);
		if (tmp != nullptr) { return tmp; }
		}
	return nullptr;
	}

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== SENSOR ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 
class SensorDirect::impl
	{
	public:
		gcroot<LHM::ISensor^> inner;
	public:
		impl(LHM::ISensor^ inner) : inner(inner) {}

	};

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== HARDWARE ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 
class Hardware::impl
	{
	public:
		gcroot<LHM::IHardware^> inner;
	public:
		impl(LHM::IHardware^ inner) : inner(inner) {}

	};

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== COMPUTER ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 

class Computer::impl
	{
	public:
		gcroot<LHM::Computer^> inner;
	public:
		impl()
			{
			inner = gcnew LHM::Computer();
			inner->IsControllerEnabled  = true;
			inner->IsCpuEnabled         = true;
			inner->IsGpuEnabled         = true;
			inner->IsMemoryEnabled      = true;
			inner->IsMotherboardEnabled = true;
			inner->IsNetworkEnabled     = false;
			inner->IsStorageEnabled     = false;

			try { inner->Open(); }
			catch (System::Exception^ e)
				{
				System::Console::WriteLine(" ===== EXCEPTION ===== ");
				System::Console::WriteLine(e->ToString());

				System::Console::WriteLine(" ===== ADDITIONAL DATA ===== ");
				for each (auto element in e->Data)
					{
					System::Console::WriteLine(element->ToString());
					}

				System::Console::WriteLine(" ===== STACK TRACE ===== ");
				System::Console::WriteLine(e->StackTrace);
				}
			}

		~impl() { inner->Close(); }
	};


// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== COMPUTER WRAPPER ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 

Computer::Computer()
	:impl_(std::make_unique<impl>())
	{}
Computer::~Computer() = default;

std::unique_ptr<Sensor> Computer::get_sensor(const std::string& identifier) const noexcept
	{
	System::String^ csidentifier = msclr::interop::marshal_as<System::String^>(identifier);
	for each (auto h in impl_->inner->Hardware)
		{
		LHM::ISensor^ tmp = ::get_sensor(h, csidentifier);
		if (tmp != nullptr)
			{
			SensorDirect::impl impl(tmp);
			return std::make_unique<SensorDirect>(impl);
			}
		}
	return {};
	}
std::unique_ptr<Hardware> Computer::get_hardware(const std::string& identifier) const noexcept
	{
	System::String^ csidentifier = msclr::interop::marshal_as<System::String^>(identifier);
	for each (auto h in impl_->inner->Hardware)
		{
		LHM::IHardware^ tmp = ::get_hardware(h, csidentifier);
		if (tmp != nullptr)
			{
			Hardware::impl impl(tmp);
			return std::make_unique<Hardware>(impl);
			}
		}
	return {};
	}


std::string _get_name(LHM::IHardware^ hardware, const std::string& identifier) noexcept
	{
	if (msclr::interop::marshal_as<std::string>(hardware->Identifier->ToString()) == identifier) { return msclr::interop::marshal_as<std::string>(hardware->Name); }

	for each (auto h in hardware->SubHardware)
		{
		auto s = _get_name(h, identifier); if (s != "") { return s; }
		}
	for each (auto s in hardware->Sensors)
		{
		if (msclr::interop::marshal_as<std::string>(s->Identifier->ToString()) == identifier) { return msclr::interop::marshal_as<std::string>(s->Name); }
		}

	return "";
	}

std::string Computer::get_name(const std::string& identifier) const noexcept
	{
	for each (auto h in impl_->inner->Hardware) 
		{
		auto s = _get_name(h, identifier); if (s != "") { return s; }
		}
	return "Name for " + identifier + " not found!";
	}

void Computer::update() noexcept
	{
	for each (auto h in impl_->inner->Hardware) { Hardware::impl impl(h); Hardware(impl).update(); }
	}

std::string Computer::to_string() const noexcept
	{
	std::string ret = "";
	for each (auto h in impl_->inner->Hardware) { Hardware::impl impl(h); ret += Hardware(impl).to_string() += "\n"; }
	return ret;
	}

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== SENSOR WRAPPER ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 

SensorDirect::SensorDirect(const SensorDirect::impl& inner) : impl_(std::make_unique<impl>(inner)) {}
SensorDirect::~SensorDirect() = default;

std::string SensorDirect::name() const noexcept { return msclr::interop::marshal_as<std::string>(impl_->inner->Name); }

std::string SensorDirect::identifier() const noexcept { return msclr::interop::marshal_as<std::string>(impl_->inner->Identifier->ToString()); }

sensor_type SensorDirect::type() const noexcept { return sensor_type_conversion(impl_->inner->SensorType); }

std::optional<float> SensorDirect::value() const noexcept
	{
	if (impl_->inner->Value.HasValue) { return std::make_optional<float>(impl_->inner->Value.Value); }
	else { return {}; }
	}

std::string SensorDirect::to_string() const noexcept { return to_string(0); }
std::string SensorDirect::to_string(uint8_t depth) const noexcept
	{
	return indent(depth) + "Sensor: " + name() + " | " + identifier() + " | " + sensor_type_to_string(type()) + "\n";
	}

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== HARDWARE WRAPPER ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 

Hardware::Hardware(const Hardware::impl & inner) : impl_(std::make_unique<impl>(inner)) {}
Hardware::~Hardware() = default;

std::string Hardware::name() const noexcept { return msclr::interop::marshal_as<std::string>(impl_->inner->Name); }

std::string Hardware::identifier() const noexcept { return msclr::interop::marshal_as<std::string>(impl_->inner->Identifier->ToString()); }

void Hardware::update() noexcept
	{
	impl_->inner->Update();
	for each (auto h in impl_->inner->SubHardware)  { Hardware::impl impl(h); Hardware(impl).update(); }
	}

std::string Hardware::to_string() const noexcept { return to_string(0); }
std::string Hardware::to_string(uint8_t depth) const noexcept
	{
	std::string ret = indent(depth) + "Hardware: " + name() + " | " + identifier() + "\n";
	for each (auto h in impl_->inner->SubHardware) { Hardware::impl impl(h); ret += Hardware(impl).to_string(depth + 1); }
	for each (auto s in impl_->inner->Sensors) { SensorDirect::impl impl(s); ret += SensorDirect(impl).to_string(depth + 1); }
	return ret;
	}

std::string SensorSum::to_string(uint8_t depth) const noexcept
	{
	return indent(depth) + s1->to_string() + "|" + s2->to_string();
	}

// ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== PRINTER ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== 

void Print()
	{
	LHM::Computer^ computer = gcnew LHM::Computer();
	computer->IsCpuEnabled = true;
	computer->IsGpuEnabled = true;
	computer->IsMemoryEnabled = true;
	computer->IsMotherboardEnabled = true;
	computer->IsControllerEnabled = true;
	computer->IsNetworkEnabled = true;
	computer->IsStorageEnabled = true;

	computer->Open();

	for each(LHM::IHardware^ hardware in computer->Hardware)
		{
		hardware->Update();
		System::Console::WriteLine(hardware->Sensors->Length);
		System::Console::WriteLine("Hardware: {0}", hardware->Name);

		for each(LHM::IHardware^ subhardware in hardware->SubHardware)
			{
			subhardware->Update();
			System::Console::WriteLine("\tSubhardware: {0}", subhardware->Name);
		
			for each(LHM::ISensor^ sensor in subhardware->Sensors)
				{
				System::Console::WriteLine("\t\tSensor: {0}, id: {1}", sensor->Name, sensor->Identifier);
				}
			}

		for each(LHM::ISensor^ sensor in hardware->Sensors)
			{
			System::Console::WriteLine("\tSensor: {0}, id: {1}", sensor->Name, sensor->Identifier);
			}
		}

	computer->Close();
	}